src/GoTrueClient.ts

```diff
@@ -822,19 +831,55 @@ export default class GoTrueClient {
     }
 
     try {
+      let syncedSession: Session | null = null
+      const refreshTokenSync = (await getItemAsync(
+        this.storage,
+        this.refreshTokenSyncStorageKey
+      )) as SyncTokenRefreshStorage | undefined
+
+      if (
+        refreshTokenSync?.refresh_token === refreshToken &&
+        refreshTokenSync?.status === 'TOKEN_REFRESHING'
+      ) {
+        await this._waitOnRefreshTokenSync()
+        const latestSession = (await getItemAsync(this.storage, this.storageKey)) as Session | null
+        if (this._isSessionValid(latestSession)) {
+          syncedSession = latestSession
+        }
+      } else if (refreshTokenSync?.status === 'TOKEN_REFRESHED') {
+        const latestSession = (await getItemAsync(this.storage, this.storageKey)) as Session | null
+        if (this._isSessionValid(latestSession)) {
+          syncedSession = latestSession
+        }
+      }
+
       this.refreshingDeferred = new Deferred<CallRefreshTokenResult>()
 
       if (!refreshToken) {
         throw new AuthSessionMissingError()
       }
-      const { data, error } = await this._refreshAccessToken(refreshToken)
-      if (error) throw error
-      if (!data.session) throw new AuthSessionMissingError()
 
-      await this._saveSession(data.session)
-      this._notifyAllSubscribers('TOKEN_REFRESHED', data.session)
+      if (!syncedSession) {
+        await setItemAsync(this.storage, this.refreshTokenSyncStorageKey, {
+          refresh_token: refreshToken,
+          status: 'TOKEN_REFRESHING',
+        })
 
-      const result = { session: data.session, error: null }
+        const { data, error } = await this._refreshAccessToken(refreshToken)
+        if (error) throw error
```

Comment:
I'm looking closer into this case when an error occurs with the request.
So far with a network disconnection, calls resolve appropriately.
I just want to make sure everything is well accounted for, so just sharing what's next on my list to validate